crates/mr_bundle/src/bundle.rs

```diff
@@ -0,0 +1,235 @@
+use crate::{
+    error::{BundleError, MrBundleResult},
+    location::Location,
+    manifest::Manifest,
+    resource::ResourceBytes,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    path::{Path, PathBuf},
+};
+
+pub type ResourceMap = HashMap<PathBuf, ResourceBytes>;
+
+/// A Manifest bundled together, optionally, with the Resources that it describes.
+/// This is meant to be serialized for standalone distribution, and deserialized
+/// by the receiver.
+///
+/// The manifest may describe locations of resources not included in the Bundle.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Bundle<M>
+where
+    M: Manifest,
+{
+    /// The manifest describing the resources that compose this bundle.
+    #[serde(bound(deserialize = "M: DeserializeOwned"))]
+    manifest: M,
+
+    /// The full or partial resource data. Each entry must correspond to one
+    /// of the Bundled Locations specified by the Manifest. Bundled Locations
+    /// are always relative paths (relative to the root_dir).
+    resources: ResourceMap,
+
+    /// Since the Manifest may contain local paths referencing unbundled files,
+    /// on the local filesystem, we must have an absolute path at runtime for
+    /// normalizing those locations.
+    ///
+    /// Passing None is a runtime assertion that the manifest contains only
+    /// absolute local paths. If this assertion fails,
+    /// **resource resolution will panic!**
+    //
+    // TODO: Represent this with types more solidly, perhaps breaking this
+    //       struct into two versions for each case.
+    #[serde(skip)]
+    root_dir: Option<PathBuf>,
+}
+
+impl<M> Bundle<M>
+where
+    M: Manifest,
+{
+    /// Creates a bundle containing a manifest and a collection of resources to
+    /// be bundled together with the manifest.
+    ///
+    /// The paths paired with each resource must correspond to the set of
+    /// `Location::Bundle`s specified in the `Manifest::location()`, or else
+    /// this is not a valid bundle.
+    ///
+    /// A base directory must also be supplied so that relative paths can be
+    /// resolved into absolute ones
+    pub fn new<R: IntoIterator<Item = (PathBuf, ResourceBytes)>>(
```

Comment:
What haapens if the dna is already installed?
Do the bytes get pulled again?

Reply:
Not sure what you mean in regards to this particular line, this is just the constructor, so you already have the bytes in-hand and you're just creating a Bundle struct from them. mr_bundle knows nothing about Holochain.